### PR TITLE
Implemented stage-based testing using DynamoDB

### PR DESCRIPTION
**Pull request by LiamBindle**
See [README.md](https://github.com/geoschem/gc-cloud-infrastructure/blob/05e75bb74db77805ef54462be9ddac1f130a3fff/benchmarks/scripts/README.md)

The basic idea is that you run one of the driver scripts in the `entrypoints/` directory. This launches a test. A test consists of stages. For example, for running a GCHP simulation there are two stages: (1) creating and setting up the run directory, and (2) running the simulation. Note that stages are run in temporary directories that are deleted after execution. A stage can upload "artifacts" (persistent data) to s3 using the `upload_artifacts` command. Before a stage is run, the artifacts from all previous stages are automatically downloaded. Note that driver scripts are reentrant, so if a test crashes because of a transient error such as a file system or network error, you can rerun the driver script and the test will pick up where it left off. 

The point of a driver script is to set the appropriate environment variables, and then run a stage. You may put all stages in a single driver script, or you can one driver script for each stage (e.g., to execute a test stages in AWS Step Functions). 

## Initial utilities for querying the database
The following command generates a table summarizing all the tests in the database:

```console
liam@compute1 $ GEOSCHEM_BENCHMARK_TABLE_NAME=geoschem_testing ./helpers/dbScanToCSV.sh
InstanceID      CreationDate  ExecStatus   Description
c5f9919.bm      2022-01-27    IN_PROGRESS  1-month GCHP benchmark simulation using 'c5f9919'
13.4.0-rc.0.bm  2022-01-27    SUCCESSFUL   1-month GCHP benchmark simulation using '13.4.0-rc.0'
```

To look at the results of a test in more depth:
```console
liam@compute1 $ GEOSCHEM_BENCHMARK_TABLE_NAME=geoschem_testing ./helpers/dbGetDetails.sh 13.4.0-rc.0.bm
{
  "Name": "SetupRunDirectory",
  "Completed": true,
  "Log": "http://s3.amazonaws.com/washu-benchmarks-cloud/13.4.0-rc.0.bm/SetupRunDirectory.txt",
  "StartTime": "2022-01-27T21:55:46+0000",
  "EndTime": "2022-01-27T22:09:47+0000",
  "PublicArtifacts": []
}
{
  "Name": "RunGCHP",
  "Completed": false,
  "Log": "",
  "StartTime": "",
  "EndTime": "",
  "PublicArtifacts": []
}
```

In the future we can extend these database queries to generate nice tables that summarize our testing, including log files and PDFs.

## Notes
- Log files and public artifacts are globally readable
- PDFs with benchmark plots are uploaded to S3 as a "public artifact" (meaning it has "everyone" read permissions so we can link to them with a URL)
- Stages log their start and end time to the database
- Stages can add custom metadata to the databased by creating a special file named `metadata.json`. The json object in this file will be added to the stage's metadata in the database. For example, this could be useful for timing info or tracking memory usage.
- Stages automatically update the tests status (PENDING, IN_PROGRESS, SUCCESSFUL, FAILED)

## Other Notes

Sorry for the bad commit messages&mdash; happy to fix them if it's a nuisance. Also, I don't feel strongly about any environment variable names (as long as they have a common prefix) or script names&mdash; happy to changes those too.

### CODE REVIEW

**Review by laestrada — APPROVED** [review]
Minor comments and a few nitpicks. Would be nice to fix some of them, but I feel comfortable with merging.

File: benchmarks/scripts/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  - `GEOSCHEM_BENCHMARK_END_DATE`: [optional]
+  - `GEOSCHEM_BENCHMARK_DURATION`: [optional]
+  - `GEOSCHEM_BENCHMARK_NUM_PROC`: [optional]
+  - `GEOSCHEM_BENCHMARK_NUM_NODES`: [optional]
+  - `GEOSCHEM_BENCHMARK_PROC_PER_NODE`: [optional]
+  - `GEOSCHEM_BENCHMARK_RESOLUTION`: [optional]
+
+The following variables must be set in any script that uses `createBenchmarkPlots.sh`:
+  - `GEOSCHEM_BENCHMARK_REFPK`: InstanceID of ref.
+  - `GEOSCHEM_BENCHMARK_DEVPK`: InstanceID of dev.
```
[the remark this answers is not in the document]
Comment (LiamBindle) [reply benchmarks/scripts/README.md]: Fixed in https://github.com/geoschem/gc-cloud-infrastructure/pull/14/commits/20339aedc3ce5fa57e225f8b51b44ac922863211

File: benchmarks/scripts/stages/setupRunDirGCHP.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+1
+${GEOSCHEM_BENCHMARK_WORKING_DIR}
+run-directory
+n
+EOF
+}
+
+function configure_run_directory() {
+    cd ${GEOSCHEM_BENCHMARK_WORKING_DIR}/run-directory
+    # set ${site_default_cores_per_node}
```
[the remark this answers is not in the document]
Comment (LiamBindle) [reply benchmarks/scripts/stages/setupRunDirGCHP.sh]: Fixed in 4e237cc

File: benchmarks/scripts/helpers/dbGetDetails.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+set -e
+set -u
+
+instance_id=$1
+
+aws dynamodb query \
```
[the remark this answers is not in the document]
Comment (LiamBindle) [reply benchmarks/scripts/helpers/dbGetDetails.sh:9]: Fixed in 4e237cc

File: benchmarks/scripts/entrypoints/benchmark-1day-wustl [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#BSUB -R 'rusage[mem=64GB] span[ptile=18]'
+#BSUB -q rvmartin
+#BSUB -W 12:00
+#BSUB -J ${GEOSCHEM_BENCHMARK_INSTANCE_ID}
+#BSUB -a 'docker(public.ecr.aws/geoschem-wustl/geoschem_benchmark:latest)'
+#BSUB -g /liam.bindle/gcst-benchmarks
+#BSUB -o /storage1/fs1/rvmartin/Active/GEOS-Chem-shared/ExternalShare/GCST/Benchmarking/Logs/%J-${GEOSCHEM_BENCHMARK_INSTANCE_ID}.txt
+#BSUB -Ne
+##BSUB -u benchmarks-cloud-aaaafk3osaonk4ruw4st4nibiq@harvard.org.slack.com 
+
```
[the remark this answers is not in the document]
Comment (LiamBindle) [reply benchmarks/scripts/entrypoints/benchmark-1day-wustl:25]: Thanks for pointing this out. I tend to agree, but right now I think it's okay to explicitly list the job properties in the WUSTL entrypoints. It's easy to edit and for now I don't know if we would benefit from an abstraction here.

**Comment by LiamBindle** [comment]
Thanks for the review @laestrada! I made a couple minor updates.

**Pull request merged**